### ISSUE TITLE
Only vertical/incorrect resolutions suggested in apps

### ISSUE THREAD

**Issue by JimmJam** [Supreeeme/xwayland-satellite#80]
Across some poorly coded applications which don't allow resolutions outside of what is initially seen by the checking what the output supports, notably some older games, only strange resolutions like 800x1280 are suggested by default.

I don't believe this to be a "primary monitor" issue, despite having a vertical display, as changing which ports the displays were plugged into did not seem to change anything, nor did changing from Niri to Sway (still using xwayland-sattelite).

![Screenshot from 2024-12-06 15-27-27](https://github.com/user-attachments/assets/26fa3e50-213d-4508-8dfb-e47313439f4e)
![Screenshot from 2024-12-06 15-18-44](https://github.com/user-attachments/assets/ec49b61e-c1a3-458a-aca9-77c963490c4c)

**Comment by Supreeeme** [comment]
What's your monitor setup?

**Comment by JimmJam** [comment]
In order of direction, HDMI-A-1 at 1080p, DP-1 at 1080p, DP-2 at 1080p, and DP-3 at 1680x1050 rotated 90 degrees. The last seems to be the monitor Satellite is locking onto for its assumed resolution.

**Comment by Supreeeme** [comment]
What games are in those screenshots? And could you post a satellite log (with `RUST_LOG=debug`) of you just starting up one of these games?

**Comment by JimmJam** [comment]
Sure thing! The games in question that were being ran were Dishonoured and Far Cry 3, here's logs of me starting sattelite, then Steam, then the game:

Dishonoured: [dishonoured.log](https://github.com/user-attachments/files/18130064/dishonoured.txt)
Far Cry 3: [farcry3.log](https://github.com/user-attachments/files/18130097/farcry3.txt)
Hitman 3: [hitman3.txt](https://github.com/user-attachments/files/18130836/hitman3.txt)

Also, I should clarify now that I've looked through the logs, the layout of my displays is as output by `niri msg outputs | grep -E 'Logical|Scale|Transform'`

```
  Logical position: 5760, 0
  Logical size: 1050x1680
  Scale: 1
  Transform: 90° counter-clockwise
  Logical position: 3840, 405
  Logical size: 1920x1080
  Scale: 1
  Transform: normal
  Logical position: 1920, 405
  Logical size: 1920x1080
  Scale: 1
  Transform: normal
  Logical position: 0, 405
  Logical size: 1920x1080
  Scale: 1
  Transform: normal
```

**Comment by Supreeeme** [comment]
Very strange - it's not immediately clear to me why it's picking the vertical monitor's resolution for the allowed in game resolutions. I'd assume a temporary workaround would be a "borderless fullscreen" option, if those games support it. Can you post the output of `xrandr`? Also, what's the behavior on Sway's Xwayland? (Not using satellite)

**Comment by JimmJam** [comment]
Output of xrandr: [xrandr.txt](https://github.com/user-attachments/files/18141517/xrandr.txt)

The behavior on Sway and Hyprland are as intended and seem to work in the same way as sattelite, at least on paper - the janky game, say, Far Cry 3, opens, forces itself into fullscreen, but correctly identifies that it happens to be on screen 1 of 3, and opens at 1920x1080. 

Sattelite seems to also correctly identify 3 monitors, but incorrectly deems the vertical one as primary, and forces that resolution into a then squished window of the same resolution of my screen, something I don't think Hyprland or Sway's implementations want to do ever (I cannot distort the window to a squished resolution on Sway's Xwayland implementation).

Borderless fullscreen doesn't work on some of these games because of how terribly they are coded and seem to ONLY want to be one of the "compatible resolutions", which is very much not this project's fault but is an unfortunate byproduct of this issue nonetheless.

**Comment by Supreeeme** [comment]
Does changing the primary monitor affect it? (`xrandr --output <output> --primary`)

**Comment by JimmJam** [comment]
Yes! That did work, Far Cry 3 is now seeing and automatically setting the resolution to the expected 1080p of DP-1 when set against it. It seems sattelite chooses its primary monitor differently to the other xwayland implementations I've tested so far.

**Comment by Supreeeme** [comment]
Hm, do the listed in game resolutions change if you set the primary monitor while the game is running too?

**Comment by JimmJam** [comment]
Yes, they do. When changing between the DP-1 and DP-3 the listed resolutions change in game.

**Comment by Supreeeme** [comment]
I added some stuff to automatically set the primary output to the [primary_output](https://github.com/Supreeeme/xwayland-satellite/tree/primary_output) branch, could you give it a try?

**Comment by JimmJam** [comment]
Just got the time to import build and test out that branch. Moved the game back and forth between the two monitors and confirmed the game recognizes the new default monitor with each switch on Niri. Awesome work!

**Comment by corngood** [comment]
> I added some stuff to automatically set the primary output to the [primary_output](https://github.com/Supreeeme/xwayland-satellite/tree/primary_output) branch

Sorry for commenting here, but I don't see a PR for 03a53b6ad7bd3d4b379f660d813ac75731d68641.

I think this change is causing bad behaviour in some cases.  Using just sway->xwayland, there doesn't appear to ever be a primary display set in Xwayland.  Using sway->waypipe->xwayland-satellite, the primary display follows the focus.

Changing the primary display sends `RRScreenChangeNotify`.  xfreerdp listens for that event and does some heavy resizing work, flickers, and picks up the resolution from the display where the mouse is.  This happens whenever I move focus between X11 apps on different outputs.  Part of it is bad behaviour in the app, but I think it's reasonable for the app to assume the screen layout isn't changing as frequently as focus.

I think it would be a good idea to make this behaviour configurable, and perhaps disable it by default.  Some compositors do appear to support the idea of a primary output:

https://wayland.app/protocols/kde-primary-output-v1

So it could be something like `--primary-output=auto|focused|none`.  I can start a PR for this if you think it's a good idea.

> It seems sattelite chooses its primary monitor differently to the other xwayland implementations I've tested so far.

There's still something going on here that I don't understand.  If sway doesn't set a primary, and xwayland-satellite didn't either (when this was tested), what explains the behaviour difference?  Just the order of the outputs?

**Comment by viceebun** [comment]
Not knowing anything about the code, my observed behavior is this; other (built in) compositor's xwayland environment provides *some* kind of concept of a primary display to say, games, as they are still able to interpret and understand "display 1" "display 2" and "display 3", acknowledging the arbitrary "display 1"  to be the primary display.

Why a display gets chosen as primary, I have *no idea.* This even comes to my own detriment sometimes, as even on Sway I've had a game randomly decide my vertical monitor is my primary one and have many, many of these same issues.

I feel because of this weird, "we have primary displays, but not really" approach that Sway and Satellite(*) have it's definitely worth having this option, but I also completely agree it should be togglable, or even configurable to be a specific display in the config, like DP-1 or such.

**Comment by corngood** [comment]
> or even configurable to be a specific display in the config, like DP-1 or such.

At least you can accomplish this by scripting `xrandr --output <output> --primary`.

> Why a display gets chosen as primary, I have no idea. This even comes to my own detriment sometimes, as even on Sway I've had a game randomly decide my vertical monitor is my primary one and have many, many of these same issues.

When I'm talking about 'primary' I'm really only talking about the literal primary in xrandr, like:

```
DP-1 connected primary 2560x1440+1920+0 (normal left inverted right x axis y axis) 700mm x 400mm
```

It sounds like you might also be experiencing heuristics that games use when there's no primary set, which could be doing just about anything.  I'll try to test out some of those games when I get a chance.

**Comment by Supreeeme** [comment]
I am hesitant to support some kind of option for primary output, just because I'd like to keep the amount of command line options to a minimum, and It Just Works :tm: behavior to the max. In this case though it does seem like there isn't a great solution to this on the Wayland side (and a spicy wayland-protocols issue as well: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/issues/179).

Before adding a command line option for it, I'd first want to see if it would be enough to just start off with the current automatic primary display behavior, and then stop it once satellite gets some xrandr event from `xrandr --primary` that it didn't trigger, and re-enable it after `xrandr --noprimary`.

Also, could you open a separate issue for this? @corngood
